src/Bedrock.Framework/Protocols/MessagePipeReader.cs

```diff
@@ -0,0 +1,224 @@
+            if (!_isCompleted)
+            {
+                // REVIEW: Use the correct value for examined
+                _reader.AdvanceTo(_consumed, _examined);
+            }
+
+            _advanced = true;
+        }
+
+        public override void CancelPendingRead()
+        {
+            _reader.CancelPendingRead();
+        }
+
+        public override void Complete(Exception exception = null)
+        {
+            _isThisCompleted = true;
+            _reader.Complete();
+        }
+
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            if (_isThisCompleted)
+            {
+                ThrowReadAfterCompleted();
+            }
+
+            if (_isCompleted)
+            {
+                _message = new ReadOnlySequence<byte>(_backlog.WrittenMemory);
+                return new ReadResult(_message, isCanceled: false, isCompleted: true);
+            }
+
+
+            if (!_advanced)
+            {
+                AdvanceTo(_message.Start);
+            }
+
+            while (true)
+            {
+                var result = await _reader.ReadAsync(cancellationToken).ConfigureAwait(false);
+
+                if (TryCreateReadResult(result, out var readResult))
+                {
+                    return readResult;
+                }
+            }
+        }
+
+        public override bool TryRead(out ReadResult readResult)
+        {
+            if (_isThisCompleted)
+            {
+                ThrowReadAfterCompleted();
+            }
+
+            if (_isCompleted)
+            {
+                if (_allExamined)
+                {
+                    readResult = default;
+                    return false;
+                }
+
+                _message = new ReadOnlySequence<byte>(_backlog.WrittenMemory);
+                readResult = new ReadResult(_message, isCanceled: false, isCompleted: true);
+                return true;
+            }
+
+            if (!_advanced)
+            {
+                AdvanceTo(_message.Start);
+            }
+
+            if (_reader.TryRead(out var result))
+            {
+                if (TryCreateReadResult(result, out readResult))
+                {
+                    return true;
+                }
+            }
+
+            if (_allExamined)
+            {
+                readResult = default;
+                return false;
+            }
+
+            _message = new ReadOnlySequence<byte>(_backlog.WrittenMemory);
+            readResult = new ReadResult(_message, _isCanceled, _isCompleted);
+            return true;
+        }
+
+        private bool TryCreateReadResult(ReadResult underlyingReadResult, out ReadResult readResult)
+        {
+            var buffer = underlyingReadResult.Buffer;
+            _isCanceled = underlyingReadResult.IsCanceled;
+            _isCompleted = underlyingReadResult.Buffer.IsEmpty && !_isCanceled;
```

Comment:
What happens if underlyingReadResult.IsCompleted is true?

Reply:
We still may have more message to parse, so we wait till we've consumed everything from the buffer and the buffer is empty..

Reply:
What if we have a partial message and are completed? We throw right?

Reply:
Yep.
But the final buffer from the underlying reader may contain multiple messages, so we don't want to return IsCompleted immediately.